### ISSUE TITLE
feature request: specify type of OS, and select image id from avaliable images

### ISSUE THREAD

**Issue by maikenp** [gc3-uzh-ch/elasticluster#502]
In the .elasticluster/config one must specify image-id for the OS to be used. It would be really nice if one instead could specify type: e.g centos6 and then elasticluster extracts list of images offered with this specification and uses that.

**Comment by maikenp** [comment]
Yes, that is true, the name or tags are not consistent with the cloud providers. One could of course give it a try by just checking for centos 6 if centos (independent on capitalisation) or el is in word, then 6 of course. And so on. Although not very elegant nor fail proof.

However, usually you are able to check what tags your provider uses. That would be enough in first instance, then I would not have to update the configuration each time a image is deprecated. So if my provider uses tags: centos_6,gold

So in configuration of elasticluster you specify
os=centos_6
os_tag=gold

Instead of
image_id=xxx


That would be more or less easy maybe? And still an improvement.

Maiken


On 9 Nov 2017, at 11:06, Riccardo Murri <notifications@github.com<mailto:notifications@github.com>> wrote:


It would be really nice if one instead could specify type: e.g centos6 and then elasticluster extracts list of images offered with this specification and uses that.

This would indeed be a very welcome addition to ElastiCluster -- but the question is: how can this be done, practically? As far as I know, there is no way to query what operating system is running on a specific image. Certainly there is none on the OpenStack cloud that I use most often.

I can imagine we could have a kind of base image search based on "tags", but then the exact spelling and meaning of tags will vary from one cloud to the other.... (E.g., one cloud might use "centos6" and another "centos-6")

I am open to suggestions, though! :-)

**Comment by riccardomurri** [comment]
On the other hand, tags are likely to be missing on many local/on-premises clouds. (They certainly are on the one I normally use.)
 
How about matching/searching image names instead?

For instance, one could have:
```
image_name=CentOS 6
image_name_match=substring
```
and this would match any image having `CentOS 6` in the name.

Or we could have:
```
image_name=CentOS *6\.[789]
image_name_match=re
```
to filter only those images whose name matches the regular expression `CentOS *6\.[789]` -- hence, "CentOS6.8", "CentOS 6.9" would all match.

I can image the following types of match should be implemented:
* `exact` -- image name must be exactly the given string (default)
* `substring` -- image name must contain the given string
* `re` -- image name must match the given regexp
* `glob` -- perform matching of image names using shell "glob" matching (i.e., `*` and `?` are wildcards); this is perhaps redundant?

**Comment by riccardomurri** [comment]
A related problem is: what if more than one image match?  

Should ElastiCluster throw an error, or select one instance according to a pre-defined criterion? The latter sounds better, but:

- there might not be enough information to implement sensible criteria (e.g., to select the "latest" image we would need to know when the image was created -- not sure that is always available)
- criteria can fail in unexpected ways (e.g., a server-side reconfiguration changes the order in which results are returned, thus making every criterion based on order/rank return a different result)

**Comment by maikenp** [comment]
Yes, exactly, if tags are not used, one could just use the image name. And then you might get several possible images. 

In that case it could anyway be very helpful to spit out the information about those images. Even if the image will not be chosen for you. In this way you could use that information (e.g. image-id) to manually change your config-file. That would save you the step of having to query it yourself either from cli or web interface.
